### ISSUE TITLE
Support `focus child|parent` command

### ISSUE THREAD

**Issue by nikitabobko** [nikitabobko/AeroSpace#5]
State of the issue: collect use cases

Please post your use cases

**Comment by nikitabobko** [comment]
In my opinion, those are very rare features even in i3. In AeroSpace, they are even less useful because AeroSpace offers normalizations that make window navigation easier

The goal is to make AeroSpace so good that `focus child | parent` won't be needed

**Comment by Diamondrubix** [comment]
it is rare, but one use case is if you have 3 terminals on top of each other with another window to the right taking up the other half of the screen you can use focus to grab all 3 terminals to move them at once.

**Comment by mik3h0** [comment]
similar to @Diamondrubix, but I also want to be able to shift a stack of windows to a different workspace or monitor. Usually this is at some point in the working day when I'm done talking to humans and want to move my e-mail client and Slack out of view, and continue looking at the IDE and web browser I have open in the same workspace.

This becomes more valuable as screen size increases as well. On a 5120x1440 screen, I'll often have 3 or 4 panes spread horizontally, some of which are horizontal stacks of 2 or more windows (usually terminals). If I want to move that stack to the edge of my screen, I have to do a bit of a dance to achieve this, usually by moving other windows in the root of the tree around the stack I actually want to move. I guess this particular use case could be handled with a "move window to far right of root" command instead, but that's still moving other windows around the leaf I wish to change.

Hope this made sense

**Comment by chr15murray** [comment]
I would like to see this too and have a similar use case to those mentioned above where I want to move a stack.

As a workaround I have these to swap a focussed stack and single window which helps a bit. It essentially moves the single window through the stack and out the otherside before returning focus to the stack.
```
# Swap Stacks with Single
alt-shift-ctrl-h = ['focus left', 'move right', 'move right', 'focus left']
alt-shift-ctrl-l = ['focus right', 'move left', 'move left', 'focus right']
```
Instead of moving the focussed stack it instead moves the single node that is in the location I want the stack in. This will of course not work as expected if you are trying to swap two stacks

**Comment by grothkopp** [comment]
I'm not 100% sure there isn't another way to solve my issue, but I want focus parent to change the layout of the root container of a populated workspace.
So when I'm on my widescreen monitor I would like to have h_tiles as the root container layout and when I'm on the small build-in screen I'd like to change it to h_accordion. If there is a way without focus parent and then calling layout, please let me know!

**Comment by TJTorola** [comment]
Not sure if there is a better alternative, but as a long time i3 user starting with aerospace I do miss `focus parent`. Perhaps that is something that will die away over time, but specifically I think I am failing to get a workflow where I can have multiple accordians next to one another in a row. With I3 I could basically manage my workflow in a single workspace because of that depth of stacks/tabs.

That being said, this feels like a rabbit hole, because the other thing that i3 has that aerospace does not to make that whole workflow much more doable was the stack/tab UI. It allowed for visual hints about what was selected and what the current state of the tree is that isn't as possible with aerospace.

**Comment by EmpathyZeroed** [comment]
My use case #1086 , very annoying to live without this when you have windows in floating mode

**Comment by nikitabobko** [comment]
Actually, one more reason why I don't want to support `focus child|parent` is because it is not easy.

1. The feature has to come with some kind of visual indication
2. It will complicate the focus handling logic for a very minor use case

The only mentioned use case in this thread is to grab a container and move it around. Well, I think `move --parent` will work just good enough for such cases https://github.com/nikitabobko/AeroSpace/issues/1190 I think it should be enough to access only one parent up. If users want to access more than 1 parent up, ~~they are sick~~ it could be supported in the future like `--parent --parent` or `--parent 2` idk

And with modes, it actually becomes possible to emulate `focus child|parent` i3 behavior:

```toml
[mode.main.binding]
  alt-a = 'mode parent' # focus parent in i3

[mode.parent.binding]
  alt-shift-h = 'move --parent left'
  alt-shift-j = 'move --parent down'
  alt-shift-k = 'move --parent up'
  alt-shift-l = 'move --parent right'
```

> That being said, this feels like a rabbit hole, because the other thing that i3 has that aerospace does not to make that whole workflow much more doable was the stack/tab UI

I hope to implement smth like `stackline` in the future https://github.com/nikitabobko/AeroSpace/issues/24#issuecomment-1872588387 (Or I will be happy if somebody contributes it)

---

I'm closing this issue in favor of https://github.com/nikitabobko/AeroSpace/issues/1190

**Comment by HekpoMaH** [comment]
I would argue it's not a minor usecase and it adds additional pressure on our human memory (we need to remember yet another  set of combinations). If there is anything that I missed during my migration to MacOS, it's how i3/sway handle containers. I could honestly live with accordions instead of tabs if I could quickly jump out and in -- e.g. **I work on smaller screens and when I have a vsplit and an accordion node on the right and I'm working somewhere in the middle of accordion and need to go left I want to do it via a simple keymap combo (focus-parent, then focus left)**.

**I would appreciate if anyone who seems themselves in the above use-case upvotes this comment**.

About how to implement focus... no idea br0, but I wonder if JankyBorders can be of any help.

To those saying go back to Linux -- not really an option because we are talking about work laptops here.

**Comment by TJTorola** [comment]
> > That being said, this feels like a rabbit hole, because the other thing that i3 has that aerospace does not to make that whole workflow much more doable was the stack/tab UI
> 
> I hope to implement smth like `stackline` in the future [#24 (comment)](https://github.com/nikitabobko/AeroSpace/issues/24#issuecomment-1872588387) (Or I will be happy if somebody contributes it)
> 
> I'm closing this issue in favor of [#1190](https://github.com/nikitabobko/AeroSpace/issues/1190)

Yeah honestly this whole feature doesn't make sense (atleast in the case of my ~sickness~ i3 workflow) unless those tabs are also implimented. The only time I've found it useful to manage parents is when those parents are a group of tabs / stacks. Perhaps if that is implimented and people start building workflows around them then this feature would make more sense.

**Comment by peauc** [comment]
I don't understand how focus parent is not integral to more people workflows. It really is essential in complexly nested trees.

**Comment by jirassimok** [comment]
My use case is much the same as HekpoMaH's; my main use case for `focus parent` is to move between different tabbed containers / accordions (my usual setup is usually two side-by side "stacks," and I'd like to be able to jump between them without changing the visible windows.

I don't really need "focus parent" specifically, but I do want better navigation around the window tree (or at least a better set of query commands that lets me implement the navigation commands myself).

**Comment by HekpoMaH** [comment]
> My use case is much the same as HekpoMaH's; my main use case for `focus parent` is to move between different tabbed containers / accordions (my usual setup is usually two side-by side "stacks," and I'd like to be able to jump between them without changing the visible windows.
> 
> I don't really need "focus parent" specifically, but I do want better navigation around the window tree (or at least a better set of query commands that lets me implement the navigation commands myself).

Trust me, if our company's linux laptops were not crashing randomly, I'd be on i3... No offence @nikitabobko
